lib/teleterm/gateway/gateway_kube_test.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gateway
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/fixtures"
+	"github.com/gravitational/teleport/lib/kube/kubeconfig"
+	"github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/gatewaytest"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func TestKubeGateway(t *testing.T) {
+	t.Parallel()
+
+	const (
+		teleportClusterName = "example.com"
+		kubeClusterName     = "example-kube-cluster"
+	)
+
+	identity := tlsca.Identity{
+		Username:          "alice",
+		Groups:            []string{"test-group"},
+		KubernetesCluster: kubeClusterName,
+	}
+	clock := clockwork.NewFakeClock()
+	proxy := mustStartMockProxyWithKubeAPI(t, identity)
+	gateway, err := New(
+		Config{
+			Clock:              clock,
+			TargetName:         kubeClusterName,
+			TargetURI:          uri.NewClusterURI(teleportClusterName).AppendKube(kubeClusterName).String(),
+			TargetUser:         identity.Username,
+			CertPath:           proxy.clientCertPath(),
+			KeyPath:            proxy.clientKeyPath(),
+			WebProxyAddr:       proxy.webProxyAddr,
+			ClusterName:        teleportClusterName,
+			CLICommandProvider: mockCLICommandProvider{},
+			RootClusterCACertPoolFunc: func(_ context.Context) (*x509.CertPool, error) {
+				return proxy.certPool(), nil
+			},
+			OnExpiredCert: func(_ context.Context, gateway *Gateway) error {
+				return trace.Wrap(gateway.ReloadCert())
+			},
+		},
+	)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		gateway.Close()
+	})
+	go gateway.Serve()
```

Comment:
Would it make sense to check if this didn't return an error?

https://github.com/gravitational/teleport/blob/5ba028bc5843db6506f4d41919221286ce60bb40/lib/teleterm/gateway/gateway_test.go#L98-L103